src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -6,193 +6,592 @@
+                    Write(chSpan);
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteQuotedString(string value)
+        {
+            WriteChar(JsonConstants.Quote);
+            // TODO: We need to handle escaping.
+            Write(value.AsSpan());
+            WriteChar(JsonConstants.Quote);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteNumber(long value)
+        {
+            int written;
+            while (!value.TryFormat(_output.Buffer, out written, JsonConstants.NumberFormat, _output.Encoder))
+                _output.Enlarge();
+
+            _output.Advance(written);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteNumber(ulong value)
+        {
+            int written;
+            while (!value.TryFormat(_output.Buffer, out written, JsonConstants.NumberFormat, _output.Encoder))
+                _output.Enlarge();
+
+            _output.Advance(written);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteDateTime(DateTime value)
+        {
+            int written;
+            while (!value.TryFormat(_output.Buffer, out written, JsonConstants.DateTimeFormat, _output.Encoder))
+                _output.Enlarge();
+
+            _output.Advance(written);
         }
 
-        private void WriteIndent()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteDateTimeOffset(DateTimeOffset value)
+        {
+            int written;
+            while (!value.TryFormat(_output.Buffer, out written, JsonConstants.DateTimeFormat, _output.Encoder))
+                _output.Enlarge();
+
+            _output.Advance(written);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteGuid(Guid value)
+        {
+            int written;
+            while (!value.TryFormat(_output.Buffer, out written, JsonConstants.GuidFormat, _output.Encoder))
+                _output.Enlarge();
+
+            _output.Advance(written);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Write(ReadOnlySpan<char> value)
+        {
+            int written;
+            while (!_output.Encoder.TryEncode(value, _output.Buffer, out int consumed, out written))
+                _output.Enlarge();
+
+            _output.Advance(written);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteAscii(ReadOnlySpan<char> values)
         {
-            if (_prettyPrint) {
-                for (int i = 0; i < _indent; i++) {
-                    _formatter.Append("    ");
+            if (UseFastUtf8)
+            {
+                int len = values.Length;
+                Span<byte> buffer = _output.Buffer;
```

Comment:
I thought about this. Though, as it is, this is not a stack-only struct right now. Caching the buffer would make this stack-only.

FYI, there are other optimizations to be had in this code. I'll create a new work item to do more performance tweaks.